### ISSUE TITLE
Products aren't sorted by position which is set in Admin panel

### ISSUE THREAD

**Issue by aleksandrakorolova** [scandipwa/scandipwa#3578]
**Description**:
Products aren't sorted by position which is set in Admin panel

**Expected behavior**:
Products are sorted by position which is set in Admin panel
![Screenshot 2021-10-19 at 10 38 33](https://user-images.githubusercontent.com/73526235/137864873-8f49401b-8a67-4f8e-8a4a-8ce8b948756d.png)

**To Reproduce**:
**Admin panel:**
1. Open Admin panel -> Catalog -> Categories -> Open Category X
2. Open 'Products in Category' block
3. Select products (A, B, C)
4. Change position for products (A-2, B-3, С-1)
5. Save
6. Open website -> Open Category X
7. Sort by Best match

**Actual result**:
Products aren't sorted by position which is set in Admin panel
![Screenshot 2021-10-19 at 10 38 26](https://user-images.githubusercontent.com/73526235/137864846-9732886a-98a6-4d14-95fe-1fdd3c77fa23.png)

**Admin panel:**
![Screenshot 2021-10-19 at 10 42 57](https://user-images.githubusercontent.com/73526235/137865443-bc5407ca-95a0-4c42-9a75-07fc527eee7e.png)

**Comment by laravieira** [comment]
**Some links that I rely on and could help:**
* [Modificaitons on Search file from ScandiPWA/CatalogGraphQl Module](https://github.com/scandipwa/catalog-graphql/commits/master/src/Model/Resolver/Products/Query/Search.php)
* [Modifications on Search file from other project](https://bitbucket.org/scandiweb/norwoodsawmills-m2/src/master/app/code/Norwood/CatalogGraphQl/Model/Resolver/Products/DataProvider/ProductSearch.php)
* [Related issue on Magento version 2.3.x on other project](https://scandiweb.atlassian.net/browse/ZOH-164)
* [Related issue on Magento GraphQl module](https://github.com/magento/magento2/issues/26817)

**Comment by aliarsalan177** [comment]
Solution is available for this now.

**Comment by carinadues** [comment]
@aliarsalan177 do you have a PR for this?

**Comment by mtamerus** [comment]
This is because of the overwrite of ProductCollectionSearchCriteriaBuilder, which removes the category filter. Magento states `Add category id to enable sorting by position`.

**Comment by aliarsalan177** [comment]
I have applied this fix to many sites and it's working fine with Scandipwa Magento 2.x

<?xml version="1.0"?>
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
    <preference for="ScandiPWA\CatalogGraphQl\Model\Resolver\Products\DataProvider\ProductSearch\ProductCollectionSearchCriteriaBuilder"
                type="Vendor\CustomPositionFixes\Model\Resolver\Products\DataProvider\ProductSearch\ProductCollectionSearchCriteriaBuilder" />
   <type name="Magento\CatalogGraphQl\Model\Resolver\Products\DataProvider\ProductSearch">
        <plugin name="vendor_category_product_search_after" type="Vendor\CustomPositionFixes\Plugin\BeforeProductSearch"/>
    </type>
        <type name="Magento\CatalogGraphQl\Model\Resolver\Products\DataProvider\Product\CompositeCollectionProcessor">
        <arguments>
            <argument name="collectionProcessors" xsi:type="array">
                <item name="joinCatalogCategoryProductForPosition" xsi:type="object">Vendor\CustomPositionFixes\Model\Resolver\Products\DataProvider\Product\CollectionProcessor\CatalogCategoryProductJoinProcessor</item>
            </argument>
        </arguments>
    </type>
</config>


These changes can do the job...!

**Comment by robertdeboer** [comment]
@aliarsalan177 If possible could you please elaborate on "this fix" - I've seen some a form of this fix but not one with 3 files. I have a site with this same exact issue. 

Is there any plan to get this fixed?

If it helps, I was able to reproduce by doing the following:

1. Create multiple categories off of the main category & assign the same products
2. Change product position and confirm products move
3. Create sub-category off of one of the categories created above and assign the same products, changing the product position in this new sub category
4. If you go to this new category page, you will see that the products are not in the correct position

In other words for me this is only happening on nested categories. If I just create a bunch of categories at the top level, the product positions are fine. It's when you start going levels deeper the position sort is not correct.
